src/object.d

```diff
@@ -3241,6 +3241,269 @@ template RTInfo(T)
     enum RTInfo = null;
 }
 
+private template Floating1(T)
+if (is(T == cfloat) || is(T == cdouble) || is(T == creal))
+{
+    // Use rt.cmath2._Ccmp instead ?
+    int compare(T f1, T f2)
+    {
+        int result;
+
+        if (f1.re < f2.re)
+            result = -1;
+        else if (f1.re > f2.re)
+            result = 1;
+        else if (f1.im < f2.im)
+            result = -1;
+        else if (f1.im > f2.im)
+            result = 1;
+        else
+            result = 0;
+        return result;
+    }
+}
+
+private template Floating1(T)
+if (is(T == float) || is(T == double) || is(T == real))
+{
+    int compare(T d1, T d2)
+    {
+        if (d1 != d1 || d2 != d2) // if either is NaN
+        {
+            if (d1 != d1)
+            {
+                if (d2 != d2)
+                    return 0;
+                return -1;
+            }
+            return 1;
+        }
+        return d1 < d2 ? -1 : (d1 > d2);
+    }
+}
+
+private template Array1(T)
+if (is(T ==  float) || is(T ==  double) || is(T ==  real) ||
+    is(T == cfloat) || is(T == cdouble) || is(T == creal))
+{
+    int compare(T[] s1, T[] s2)
+    {
+        auto len = s1.length;
+        if (s2.length < len)
+            len = s2.length;
+
+        foreach (const u; 0 .. len)
+        {
+            if (int c = Floating1!T.compare(s1[u], s2[u]))
+                return c;
+        }
+        return s1.length < s2.length ? -1 : (s1.length > s2.length);
+    }
+}
+
+int _adCmpT(T1, T2)(T1[] s1, T2[] s2)
+{
+    import core.internal.traits : Unqual;
+    alias U1 = Unqual!T1;
+    alias U2 = Unqual!T2;
+
+    // objects
+    // Old code went to object.d/TypeInfo_array.compare and then to
+    // object.d/TypeInfo_Class.compare
+    // which combined result in the same code as below.
+    static if (is(U1 : Object) && is(U2 : Object))
+    {
+        auto c  = cast(sizediff_t)(s1.length - s2.length);
+        if (c == 0)
+        {
+            for (size_t u = 0; u < s1.length; u++)
+            {
+                auto o1 = () @trusted { return s1.ptr[u]; }();
+                auto o2 = () @trusted { return s2.ptr[u]; }();
```

Comment:
Without optimization, the compiler still generates a call to a function:
```
int test1(Object[] a, Object[] b) {
    static @trusted ref R at(R)(R[] r, size_t i) { return r.ptr[i]; }
	auto o1 = at(a, 0);
	auto o2 = at(b, 0);
    return 0;
}

int test2(Object[] a, Object[] b) {
	auto e1 = () @trusted { return a.ptr[0]; }();
	auto e2 = () @trusted { return b.ptr[0]; }();
  
  	if (e1 != e2) {}
  	return 0;
}
```

```
---
call    pure nothrow ref @nogc @trusted Object example.test1(Object[], Object[]).at!(Object).at(Object[], ulong)
---
```

From https://godbolt.org/g/IykECY